### PR TITLE
[WIP] Fix subscription urls

### PR DESCRIPTION
**Pull request by Edzelopez**
Overview
----------------------------------------
This is an initial attempt to fix the errors when cancelling and updating billing details for a recurring contribution using self serve links. [#127](https://lab.civicrm.org/dev/financial/-/issues/127)

Before
----------------------------------------
1. Trying to cancel a recurring payment using the self serve link results in an error:
InvalidArgumentException: "processorID field has max length of 255"

#0 /xxx/civicrm/CRM/Contribute/Form/CancelSubscription.php(217): Civi\Payment\PropertyBag->setRecurProcessorID(NULL)

This was due to the absence of the processor_id in the civicrm_contribution_recur table.

2. Trying to update billing details, gives the following error:
Exception: "This system is unable to perform self-service updates to credit cards. Please contact the administrator of this site."

#0 /xxx/civicrm/CRM/Contribute/Form/UpdateBilling.php(207): CRM_Core_Payment_iATSService->updateSubscriptionBillingInfo(NULL, (Array:21))

This was due to the absence of the crid(recur_id) param in the URL.

After
----------------------------------------
The links now function as expected, allowing one to cancel or update the billing details for a recurring contribution.

Comments
----------------------------------------
The fix around passing the extra recur_id param may likely need to be expanded to handle the various other calls to this function from different instances. Hence, marked this as WIP, to open discussions for the same.

### CODE REVIEW

**Comment by mattwire** [comment]
I would prefer to see these moved to a token implementation rather than being fixed via the smarty route. We've made some good progress on tokens recently.

I'd suggest:
1. If we can get https://github.com/civicrm/civicrm-core/pull/16983 reviewed and merged then it becomes simpler to implement a contributionRecur tokenProcessor class.
2. You can more or less copy `CRM_Activity_Tokens` to create `CRM_Contribute_Recurtokens` (or similar).
3. Trigger tokenProcessor when sending the message template so we parse contributionrecur tokens. (This also gives us a path to deprecate the older token methods later). I can help with this part.
3. Pass the recur ID through with message template params so that it's picked up by tokenProcessor.

**Comment by mattwire** [comment]
Additional benefit would be that we'd get recurring contribution tokens that we don't currently have.

**Comment by JoeMurray** [comment]
It seems like you're asking for a fair bit of work here, @mattwire . @seamuslee001 is going to take a look too and provide feedback. We've got a limit budget and even more limited timeframe for completing this.

**Comment by mattwire** [comment]
@JoeMurray I know the feeling :-) In that case let's fix the bug but see if we can do anything to make future cleanup easier. Just improving the comments on the params to `subscriptionURL` would certainly help.

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 1]
```diff
@@ -1088,7 +1088,7 @@ public static function processRecurringContribution(&$form, &$params, $contactID
     }
     $recurParams['invoice_id'] = $params['invoiceID'] ?? NULL;
     $recurParams['contribution_status_id'] = CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'Pending');
-    $recurParams['payment_processor_id'] = $params['payment_processor_id'] ?? NULL;
+    $recurParams['payment_processor_id'] = $recurParams['processor_id'] = $params['payment_processor_id'] ?? NULL;
```
Comment (seamuslee001) [inline CRM/Contribute/Form/Contribution/Confirm.php:1091]: I'm not 100% sure this is correct, 

The processor_id is meant to be the unique token or similar that is used when charging the payment. The payment_processor_id = the id of the PaymentProcessor that is used from civicrm_payment_processor table

Comment (Edzelopez) [reply CRM/Contribute/Form/Contribution/Confirm.php:1091]: Makes sense. I figured the processor_id == payment_processor_id, which is not true.

**Comment by eileenmcnaughton** [comment]
I'm pretty sure this was otherwise fixed as a regression a while back

"#0 /xxx/civicrm/CRM/Contribute/Form/CancelSubscription.php(217): Civi\Payment\PropertyBag->setRecurProcessorID(NULL)"

- now treats NULL as acceptable

**Pull request closed without merging**